Fix DDP bug in single process multiple device use cases

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36503 [WIP] Fix DDP bug in single process multiple device use cases**

Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

Commit:
Fix DDP bug in single process multiple device use cases

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -314,6 +314,20 @@ def _ddp_init_helper(self):
+                for p in ps:
+                    yield p
+
+            ms = m.modules() if recurse else [m]
```
what is this for? looks like it is not used?

mrshenli replied:
my bad will update.

zhaojuanmao reviewed (commented):
sorry, accidentally clicked approve

zhaojuanmao commented on test/distributed/test_c10d.py:
```diff
@@ -1871,6 +1871,89 @@ def forward(self, x):
 
 
+@unittest.skipIf(TEST_WITH_TSAN, "TSAN is not fork-safe since we're forking in a multi-threaded environment")
+class DistributedDataParallelSingleProcessTest(TestCase):
```
it has distributeddataparallel test in test_distributed.py,  do you think whether it is better  to put tests in test_distributed.py?

mrshenli replied:
Sure, let me move it. Would I be correct if I assume we plan to move all DDP related tests to `test_distributed.py`?

zhaojuanmao replied:
yeah, I feel it is better to put ddp related tests in test_distributed.py, how do you think?

mrshenli replied:
Hey @zhaojuanmao, I compared `test_distributed.py` and `test_c10d.py`. If we plan to make them `test_ddp.py` and `test_c10d.py` in the future, I agree that this test better fits in `test_ddp.py`. But for now, it looks it fits more in `test_c10d.py`, as `DistributedDataParallelSingleProcessTest` is the single-process version of `DistributedDataParallelTest ` in `test_c10d.py` immediately after this test. And currently, a considerable part of `test_distributed.py` are testing APIs on c10d backend. Shall we keep this test in here for now, and we can certainly move it when we refactor the test file structures.

Commit:
Update on "Fix DDP bug in single process multiple device use cases"




[ghstack-poisoned]

Commit:
Update on "Fix DDP bug in single process multiple device use cases"




[ghstack-poisoned]

Commit:
Update on "Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

Commit:
Update on "Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

Commit:
Update on "Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

Commit:
Don't crash when BACKEND and TEMP_DIR env var are missing if tests don't need them

[ghstack-poisoned]

Commit:
Update on "[WIP] Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

Commit:
Update base for Update on "[WIP] Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

Commit:
Update on "[WIP] Fix DDP bug in single process multiple device use cases"


Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
sounds good, I will think about how to refactor the test structures as well.

thanks for fixing it.
